packages/@lwc/ssr-compiler/src/compile-js/generate-markup.ts

```diff
@@ -62,7 +62,7 @@ const bGenerateMarkup = esTemplate`
             instance.connectedCallback();
             __mutationTracker.disable(instance);
         }
-        const tmplFn = ${isIdentOrRenderCall} ?? ${/*component class*/ 3}[__SYMBOL__DEFAULT_TEMPLATE] ?? __fallbackTmpl;
+        const tmplFn = instance.render?.() ?? ${isIdentOrRenderCall} ?? ${/*component class*/ 3}[__SYMBOL__DEFAULT_TEMPLATE] ?? __fallbackTmpl;
```

Comment:
This is working because basically we are checking if our instance has a `render()` or if any of its superclasses does, and then deferring to that. Then it will fall back to the `__SYMBOL__DEFAULT_TEMPLATE` which is only set if the default template exists, so again this works correctly with inheritance.

This actually seems to match `engine-server`'s behavior, which is indeed to defer to `render()` and fall back from there.

I think thought that we can simplify this code a lot. First off, let's add a variable name to make it clear what `isIdentOrRenderCall` is doing:

```suggestion
        const tmplFn = instance.render?.() ?? ${/* renderCall */ isIdentOrRenderCall} ?? ${/*component class*/ 3}[__SYMBOL__DEFAULT_TEMPLATE] ?? __fallbackTmpl;
```

`renderCall` is doing:

```ts
    const renderCall = hasRenderMethod
        ? (b.callExpression(
              b.memberExpression(b.identifier('instance'), b.identifier('render')),
              []
          ) as RenderCallExpression)
        : tmplVar;
```

So clearly it doesn't need to actually check `hasRenderMethod` anymore (anyway it can't know – it might be inherited from a superclass/mixin). So if you remove it and just put `tmplVar` then all the tests still pass.

Here are some more tests to gain confidence that we are doing things correctly: https://github.com/salesforce/lwc/pull/5065

Your PR passes all these tests too. 🙂